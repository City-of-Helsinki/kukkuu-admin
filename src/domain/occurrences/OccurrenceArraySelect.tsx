import React, { ChangeEvent } from 'react';
import {
  SelectArrayInput,
  ReferenceArrayInput,
  useTranslate,
  useInput,
} from 'react-admin';
import { useFormContext } from 'react-hook-form';

import { Occurrence_occurrence as Occurrence } from '../../api/generatedTypes/Occurrence';
import { toShortDateTimeString } from '../../common/utils';

const AllChoice = ({
  children,
  choices,
  allChoiceLabel,
  getChoices,
  ...rest
}: any) => {
  return React.cloneElement(children, {
    ...rest,
    choices: [{ id: 'all', name: allChoiceLabel }, ...getChoices(choices)],
  });
};

export function getFilters(
  eventId?: string
): Record<string, string> | undefined {
  if (!eventId) {
    return;
  }

  return {
    eventId,
  };
}

export const getChoices = (records: Occurrence[]) => {
  return records.map(({ id, time }) => ({
    id,
    name: toShortDateTimeString(new Date(time)),
  }));
};

export const getValues = (previousValues: string[], nextValues: string[]) => {
  const wasAllSelected = previousValues.includes('all');
  const willHaveOtherValueThanAll = nextValues.length > 0;
  const willHaveNoValue = nextValues.length === 0;
  const allWillBeAdded =
    !previousValues.includes('all') && nextValues.includes('all');

  if (wasAllSelected && willHaveOtherValueThanAll) {
    return nextValues.filter((value: string) => value !== 'all');
  } else if (willHaveNoValue || allWillBeAdded) {
    return ['all'];
  } else {
    return nextValues;
  }
};

type ReferenceArrayInputProps = Parameters<typeof ReferenceArrayInput>[0];
type Props = Omit<ReferenceArrayInputProps, 'children' | 'reference'> & {
  source: string;
  eventId?: string;
  allText: string;
};

const OccurrenceArraySelect = ({ eventId, allText, ...rest }: Props) => {
  const filter = getFilters(eventId);
  const {
    field: { value, name },
  } = useInput({ source: rest.source });
  const form = useFormContext();
  const t = useTranslate();

  // By default the multi select component does not explicitly have an
  // "all" option. This custom handler implements behavior for the all
  // option.
  // - when the user deselects all choices, all is selected
  // - when the user selects some (other) option (than all), "all"
  //   selection is removed
  // - when the user selects "all", all other choices are removed
  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const previousValues = value;
    // The value is actually an array because we are using a
    // multi select
    const nextValues = e.target.value as unknown as string[];
    const values = getValues(previousValues, nextValues);

    form.setValue(name, values);
  };

  return (
    <ReferenceArrayInput
      {...rest}
      resource="occurrences"
      reference="occurrences"
      filter={filter}
      onChange={handleChange}
    >
      <AllChoice allChoiceLabel={t(allText)} getChoices={getChoices}>
        <SelectArrayInput />
      </AllChoice>
    </ReferenceArrayInput>
  );
};

export default OccurrenceArraySelect;
