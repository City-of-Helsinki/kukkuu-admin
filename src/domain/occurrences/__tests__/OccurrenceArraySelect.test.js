import React from 'react';
import { render } from '@testing-library/react';
import { TestContext, SimpleForm } from 'react-admin';

import OccurrenceArraySelect, {
  getChoices,
  getFilters,
  getValues,
} from '../OccurrenceArraySelect';

describe('<OccurrenceArraySelect />', () => {
  it('should render without errors', () => {
    jest.spyOn(global.console, 'error').mockImplementationOnce(() => null);

    expect(
      render(
        <TestContext
          initialState={{ admin: { resources: { occurrences: { data: {} } } } }}
        >
          <SimpleForm>
            <OccurrenceArraySelect
              allText="allText"
              eventId="1"
              source="event"
            />
          </SimpleForm>
        </TestContext>
      )
    ).toMatchSnapshot();
  });
});

describe('OccurrenceArraySelect helpers', () => {
  describe('getFilters', () => {
    it('should return undefined if there is no event id', () => {
      expect(getFilters()).toBeUndefined();
    });

    it('otherwise it should return filters', () => {
      expect(getFilters('123')).toEqual({
        eventId: '123',
      });
    });
  });

  describe('getChoices', () => {
    it('should make choices out of occurrences', () => {
      const occurrences = [
        {
          id: '123',
          time: '2020-10-7',
        },
        {
          id: '12',
          time: '2020-11-7',
        },
      ];
      const choices = getChoices(occurrences);

      expect(choices.length).toEqual(occurrences.length);
      expect(choices[0]).toMatchInlineSnapshot(`
        Object {
          "id": "123",
          "name": "7.10.2020, 00.00",
        }
      `);
    });
  });

  describe('getValues', () => {
    it('should remove all from selected values if some other option is selected', () => {
      expect(getValues(['all'], ['other options']).includes('all')).toBe(false);
    });

    it('should only return all if all is selected', () => {
      expect(getValues(['value 1', 'value 1'], ['all'])).toEqual(['all']);
    });

    it('should return all if nothing is selected', () => {
      expect(getValues(['value 1', 'value 1'], [])).toEqual(['all']);
    });

    it('otherwise it should return next values', () => {
      const nextValues = ['value 1', 'value 2'];

      expect(getValues(['value 1'], nextValues)).toEqual(nextValues);
    });
  });
});
