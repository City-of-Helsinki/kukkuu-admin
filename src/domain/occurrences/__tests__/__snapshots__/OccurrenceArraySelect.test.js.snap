// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OccurrenceArraySelect /> should render without errors 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="simple-form"
      >
        <div
          class="MuiCardContent-root RaCardContentInner-root-1"
        >
          <div
            class="ra-input ra-input-event"
          >
            <div
              class="MuiFormControl-root RaFormInput-input-2 MuiFormControl-marginDense"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root RaLabeled-label-3 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense"
                data-shrink="true"
              >
                <span>
                  resources.occurrences.fields.event
                </span>
              </label>
              <div
                class="RaLabeled-value-4"
              >
                <div
                  class="MuiLinearProgress-root MuiLinearProgress-colorPrimary RaLinearProgress-root-5 MuiLinearProgress-indeterminate"
                  resource="occurrences"
                  role="progressbar"
                >
                  <div
                    class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
                  />
                  <div
                    class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiToolbar-root MuiToolbar-regular RaToolbar-toolbar-6 RaToolbar-mobileToolbar-8 MuiToolbar-gutters"
          role="toolbar"
        >
          <div
            class="RaToolbar-defaultToolbar-9"
          >
            <button
              aria-label="ra.action.save"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained RaSaveButton-button-11 MuiButton-containedPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root RaSaveButton-leftIcon-12 RaSaveButton-icon-13"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                  />
                </svg>
                ra.action.save
              </span>
            </button>
          </div>
        </div>
        <div
          class="RaToolbar-spacer-10"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="simple-form"
    >
      <div
        class="MuiCardContent-root RaCardContentInner-root-1"
      >
        <div
          class="ra-input ra-input-event"
        >
          <div
            class="MuiFormControl-root RaFormInput-input-2 MuiFormControl-marginDense"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root RaLabeled-label-3 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-marginDense"
              data-shrink="true"
            >
              <span>
                resources.occurrences.fields.event
              </span>
            </label>
            <div
              class="RaLabeled-value-4"
            >
              <div
                class="MuiLinearProgress-root MuiLinearProgress-colorPrimary RaLinearProgress-root-5 MuiLinearProgress-indeterminate"
                resource="occurrences"
                role="progressbar"
              >
                <div
                  class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
                />
                <div
                  class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiToolbar-root MuiToolbar-regular RaToolbar-toolbar-6 RaToolbar-mobileToolbar-8 MuiToolbar-gutters"
        role="toolbar"
      >
        <div
          class="RaToolbar-defaultToolbar-9"
        >
          <button
            aria-label="ra.action.save"
            class="MuiButtonBase-root MuiButton-root MuiButton-contained RaSaveButton-button-11 MuiButton-containedPrimary Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root RaSaveButton-leftIcon-12 RaSaveButton-icon-13"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
                />
              </svg>
              ra.action.save
            </span>
          </button>
        </div>
      </div>
      <div
        class="RaToolbar-spacer-10"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
